doc/sphinx/source/n3fit/hyperopt.rst

```diff
@@ -386,3 +386,37 @@ To achieve this, you can use the ``--restart`` option within the ``n3fit`` comma
 
 The above command example is effective when the number of saved trials in the ``test_run/nnfit/replica_1/tries.pkl`` is
 less than ``20``. If there are ``20`` or more saved trials, ``n3fit`` will simply terminate, displaying the best results.
+
+
+Running hyperoptimizations in parallel with MongoDB
+---------------------------------------------------
+
+In NNPDF, you can effectively run hyperoptimization experiments in parallel using `MongoDB <https://www.mongodb.com>`_.
+This functionality is provided by the :class:`~n3fit.hyper_optimization.mongofiletrials.MongoFileTrials` class,
+which extends the capabilities of `hyperopt <https://github.com/hyperopt/hyperopt>`_'s `MongoTrials` and enables the
+simultaneous evaluation of multiple trials.
+
+To run a parallelized hyperopt search, use the following command:
+
+.. code-block:: bash
+
+  n3fit hyper-quickcard.yml 1 -r N_replicas --hyperopt N_trials --parallel-hyperopt --num-mongo-workers N
+
+Here, ``N`` represents the number of MongoDB workers you wish to launch in parallel.
+Each mongo worker handles one trial in Hyperopt. So, launching more workers allows for the simultaneous calculation of a greater number of trials.
+Note that there is no need to manually launch MongoDB databases or mongo workers prior to using ``n3fit``,
+as the ``mongod`` and ``hyperopt-mongo-worker`` commands are automatically executed
+by :meth:`~n3fit.hyper_optimization.mongofiletrials.MongodRunner.start` and
+:meth:`~n3fit.hyper_optimization.mongofiletrials.MongoFileTrials.start_mongo_workers` methods, respectivelly.
+By default, the ``host`` and ``port`` arguments are set to ``localhost`` and ``27017``, while the database is named ``hyperopt-db``.
+If necessary, you can modify these settings using the ``n3fit --db-host`` , ``n3fit --db-port`` and ``n3fit --db-name`` options.
+
+To resume a hyperopt experiment, add the ``--restart`` option to the ``n3fit`` command:
+
+.. code-block:: bash
+
+  n3fit hyper-quickcard.yml 1 -r N_replicas --hyperopt N_trials --parallel-hyperopt --num-mongo-workers N --restart
+
+Note that, unlike in serial execution, parallel hyperoptimization runs do not generate ``tries.pkl`` files.
+Instead, MongoDB databases are saved as ``hyperopt-db.tar.gz`` files inside ``replica_path`` directory.
+These are conveniently extracted for reuse in restart runs.
```

Comment:
I am actually wondering if we should either (a) store the history of the database inside the `fit_path` or (b) append to `hyperopt-db` the name of the runcard? The reason is that if one runs different hyperopts with different runcard (with for example different dataset selections) inside a given directory then everything will mix up.

Reply:
Yes, you are certainly right. we could leave the name of the tar file as general as `hyperopt-db.tar.gz` inside the `fit_path` but compress instead the database name + runcard name ? What do you think ?

Reply:
A per the discussion, please do whichever is easier (implementation-wise). Option (a) would be nicer but if it turns out to introduce more (unneeded) complications then we should go for (b).

Reply:
We implemented option b. It was failing the hyperopt tests locally, but for seemingly unrelated reasons (I had to install pdfflow?). Let's see if it passes the CI.

Reply:
The CI/CD will now pass!!

Reply:
Perfect! Thanks